include/wabt/ir.h

```diff
@@ -255,12 +255,24 @@ struct FuncSignature {
   TypeVector param_types;
   TypeVector result_types;
 
-  // Some types can have names, for example (ref $foo) has type $foo.
-  // So to use this type we need to translate its name into
-  // a proper index from the module type section.
-  // This is the mapping from parameter/result index to its name.
-  std::unordered_map<uint32_t, std::string> param_type_names;
-  std::unordered_map<uint32_t, std::string> result_type_names;
+  // Reference types can have names, for example (ref $foo) represents
+  // a type which name is $foo. These types are translated to their
+  // corresponding index after the parsing is completed. The position
+  // and names of these reference types are stored in a ReferenceVars
+  // vector. The names are stored as variables, because the error
+  // message construction requires a location when a name is not found.
+  struct ReferenceVar {
+    ReferenceVar(uint32_t index, Var var)
+      : index(index), var(var) {}
+
+    uint32_t index;
+    Var var;
+  };
+
+  typedef std::vector<ReferenceVar> ReferenceVars;
+
+  ReferenceVars param_type_vars;
+  ReferenceVars result_type_vars;
```

Comment:
I guess at some point it might make sense to move this parser-specific metadata out of the IR representation?  But it seems like this change just adds to the existing pattern?

Reply:
I think yes. Yes, the param_type_names/result_type_names was there, I just changed them to another structure.

There are many text parser related stuff in the code, since declaration order is not fixed in text parsing, sometimes things are declared after use. Maybe they could be hidden behind smart pointers, so they have less memory overhead when not used anymore.

Or use class inheritance, and reallocate the signature without parser data after parsing is completed.

Ok I will think about it.